### ISSUE TITLE
Problem: Watched folders for automatic import/filesystem based import not included in analyzer.

### ISSUE THREAD

**Issue by Robbt** [libretime/libretime#70]
I believe this functionality is missing from analyzer but the ability to bulk import tracks via FTP or the filesystem is pretty useful especially for people starting to use LibreTime from a different automation system. This was called the watched folders feature.

**Comment by hairmare** [comment]
There is the [ftp upload hook](https://github.com/LibreTime/libretime/blob/a583d8fa1e9ceaa8d9a37a9043262a8452914869/python_apps/airtime_analyzer/tools/ftp-upload-hook.sh) that might replace parts of the watched folder feature and I also found `airtime-import` in utils which seems to be somewhat related.

**Comment by hairmare** [comment]
I have the UI changes prepared on https://github.com/LibreTime/libretime/compare/master...radiorabe:feature/restore-media-folders-ui.

The calls to media-monitor that need to get changed to analyzer are on [airtime_mvc/application/models/MusicDir.php#L296](https://github.com/LibreTime/libretime/blob/cb4ad730fc9ccb342bb46a26c5074800957494e4/airtime_mvc/application/models/MusicDir.php#L296) and [#L383](https://github.com/LibreTime/libretime/blob/cb4ad730fc9ccb342bb46a26c5074800957494e4/airtime_mvc/application/models/MusicDir.php#L383).

I started looking at media-monitor but didn't get very far, may re-implementing this in analyzer greenfield style makes more sense.

**Comment by samcarmex** [comment]
Just wondering if there's any update on this? Monitoring a network mounted folder is now impossible without the airtime-import function working. We store all our audio content on a separate server (same local network though) and simply 'import' the files into Airtime without having to copy them to a second folder. This can't be done with Libretime (AFAIK).

Are there any plans to implement a true watched folder functionality or will that not be a priority?

**Comment by Robbt** [comment]
Work has been done under #111 but I'm not sure what state it is in. Either the original contributor @HaJoHe will need to finish the work or someone else will need to do whatever is left that needs to be done to bring this functionality back. I wouldn't say it's not a priority but it hasn't been on the top of my todo list at the moment.

**Comment by HaJoHe** [comment]
Hi,
all works fine for me. So, if there are any modifications needed, I'll try to do it.

**Comment by Robbt** [comment]
Ok, I had an installation issue w/ your PR. I was getting permission errors writing to the egg file. I'm trying it again from a fresh vagrant box and see if I can get it to work. It has been a regularly requested feature and we should commit the PR if it works out of the box.

If I remember correctly the only major objection is that you are writing directly to the database vs. going through the API. I'm not sure if this will cause issues down the line but I'm not sure any hooks happen that alter the files on their way through the system anyways.

**Comment by samcarmex** [comment]
Just wondering if any further work has been done on this issue? Watched folder functionality still doesn't exist AFAIK and is the single reason I'm still clutching onto my outdated Airtime install. Really wanting to upgrade to LibreTime but doesn't seem like it's there yet?

**Comment by Robbt** [comment]
I started to look into it and test it. It appears at first glance that there is an issue with the python magic module since she switched after this was coded. I agree that this is probably the highest priority for the project at the moment as it makes adopting LibreTime rather time consuming.  If I have the time I plan on doing whatever I can to get this finished.

**Comment by hairmare** [comment]
IMO this is one the larger blockers since some stations have large audio archives (>1TiB) that would get duplicated if we have users upload them to LibreTime.

**Comment by Robbt** [comment]
Yeah I think that we need to really think about how he we do it if we want to keep it as it adds a lot of potential for failure if the files are remotely hosted.

**Comment by hairmare** [comment]
If I understood the old code correctly syncing the watched folders to the database was done by cron regularly, maybe looking into inotify and having a libretime-watch service could also make sense.

**Comment by Robbt** [comment]
Yeah I was using pynotify which uses inotify for the LibreTime-import service I wrote but being inotify based it won't support remote filesystems such as sshfs. The biggest challenge is figuring out the minimalist changes to support imports that don't change the filepath while using the rest media API.

**Comment by hairmare** [comment]
We certainly need to extend the REST API to be able to do this properly. I looked into this ages ago and it's clearly a cross cutting concern that needs lots of work.

**Comment by Robbt** [comment]
Yeah, so this is why I'm not sure we want to include "watched folders" before we start the rewrite of the backend. I tried to cludge my way through modifying @HaJoHe PR in a separate branch but I found the process of adding parameters on multiple PHP functions that were chained together to simply say "don't copy this file but use the filesystem path being passed" to be problematic. I could go back and try it again or else write a whole separate API call for "pre-existing files" that attempts to simplify it. Or we could just do something with a filesystem import service that allows people to add tracks via the filesystem and moves them into LibreTime. Either way #508 should be fixed so that the problem files don't clog the import directory.

My personal take on this is if people have a well organized library of audio that they want to maintain they should do this on a separate computer from their LibreTime server and copy the files into LibreTime to be imported and stored separately on the LT file system. We probably need to add documentation for how to do this via FTP or SSH or via flash drive what not.

The only time there would be duplication of files is if they installed LT on a AT system where they had maintained a separate directory that they watched and then they install LT and the "watched folders" tracks get into an unknown state. Actually the question of what happens to "watched folders" and the tracks on a upgraded LT is a good question and probably needs to be addressed especially if we decide to not support this feature.

But I think for a lot of people being able to upload tracks via the file system will solve the problem of requiring them to manually upload a bunch of tracks through the web interface. Some kind of web-export/mass download would also be useful because I know people using airtime.pro have no way of getting their data out and transitioning to LT other than manually downloading everything.

**Comment by Robbt** [comment]
Well it would be helpful if you didn't characterize a code review/critique as a personal attack Everyone communicates differently and it is important for the sanity of the project to assume good will when someone is critical or questioning of a contribution. In my opinion we don't have a situation where people are being personally attacked or insulted so it'd be great if you could try to not take things personally. 

I think the core disagreement that has happened and you reference is more of a contradiction between two different coding philosophies, the quick hack vs. the optimal solution. In my opinion as a project we are striving for the optimal while wading through the muck of technical debt that makes up our codebase. I think we should provide some kind of outlet for the kind of one-off scripts and alternative RSS importers that you have developed. If you want to test them out in your fork and get them working then perhaps we can bring them back later. 

But I'm also of the mindset that we should try to do the best we can and verify that our code is correct before we commit it. In an ideal world every piece of code contributed would have corresponding documentation and tests to verify its validity but we have a very bloated codebase already that is lacking in these regards.

On the other hand a lot of your scripts might be helpful if added to something like a utils/contrib folder with accompanying documentation so people could use them as they needed.

I wrote a python service #514 for the import script which probably does essentially the same thing as your script but just conforms more to our current service orientated paradigm and once I add documentation and we decide upon whether we want to support "watched folders" or copying existing files vs. just files that are copied into this directory then we can merge it.

**Comment by JohnnyC1951** [comment]
I will not repeat here the 'critique', it was not at all polite. I had merely made an already posted solution work correctly, by parsing airtime.conf correctly so it actually returned the username & password (it was returning nulls) and executing the hook as a service, with the correct credentials, in the same way as, I guess, you realized it in python. That was the point of the little 20 line stand-alone sample 'how to use the hook'. Then the roof fell in. I will not apologize for using external code, on the plus side, we gained useful additional functionality for 7 years on AT and the LT, with some external add-ons. such as the selective remote provisioning of dozens of LT ARM boxes for use 'in-store' radios, basically using rsysnc, RSS and that little loop. They are now being turned into Open Broadcaster clients by the old team. A good move for them as they can now add 'in-store' signage. Without my little cron jobs, we would still be waiting.

**Comment by JohnnyC1951** [comment]
Oh, it was stinging and personal. I know the difference. Signing out. I am easy to find. The main fork has just lost an asset. I did give it a fair second go,

**Comment by Robbt** [comment]
@JohnnyC1951 I'm sorry you feel that way. I think you are referring to the script posted here - #181 - and I don't see anything overtly personal or attacking in @hairmare's response:

> Your description sounds like import_a_track.sh does exactly what ftp-upload-hook.sh should do (and calling it for multiple files is a find one-liner in my book). If there is an issue with the upload hook, please describe it properly so we can get it fixed.

I can see how that response could be seen as dismissive of the contribution but I don't think it should be taken as an personal attack. I found the script posted there to be useful but like I mentioned above figuring out a way to include scripts and hacks that aren't part of the web UI or core of the project. But yeah, I still stand behind the point I made above. It is very easy to perceive malice when reading the written word. Like I said above we should try to be respectful of each other and work together. The best thing this project has going for it is community and constantly jumping in and out and getting upset is not helpful towards the vibe. 

I know the AT forum had some contributors who were just venomous at times and slung insults and bizarre metaphors in a negative fashion but I don't see this type of negativity being directed towards you on here. If others see things differently they can chime in. We should be trying to build a cooperative project and like I said above I think the scripts that you use are useful but should be added to a part of the project where people can find them and use them. We don't want the github issue queue to become a repository for hacks etc, those could be posted on the forum but I'd prefer the ones that are useful to be added to utils/contrib or wherever with at least minimal text documentation. If you want to contribute them via PR.

**Comment by JohnnyC1951** [comment]
This was the watershed. Yes is does seem like what ftp-upload-hook.sh should do. The key point is....... That didn't work!

**Comment by Robbt** [comment]
Which is exactly where I see the problem occurring. I did find ftp-upload-hook.sh to work when I finally tried to figure out what was happening. @ghost got offended and immediately projected an antagonistic attitude on what I think was a reasonable question. Honestly we all have different personalities here and we might rub each other the wrong way at times or perceive hostility that wasn't intended. It isn't helpful if you are going to now go around being antagonistic and passive aggressive because you feel like your contributions weren't appreciated as much as they should have been.

 Can you try to let bygones be bygones and not bring up perceived snubs of the past so that we can all try to foster a cooperative atmosphere and not one rife with tension. I am not trying to dismiss your feelings as invalid, I understand that everyone perceives reality in a different way but it is important for the project as a whole that we try and treat each other with respect even if we have different opinions about code. This is essential if this project is going to survive and reach its full potential.

**Comment by JohnnyC1951** [comment]
He is in Canada reprogramming the Arm Boxes. I agreed with him.
If you look over my stuff from today... I am not sure it was me who actually blew.

**Comment by JohnnyC1951** [comment]
Are we all done?

**Comment by JohnnyC1951** [comment]
OK we are all done.

**Comment by Robbt** [comment]
So this has been targeted for the beta but in reality we need to make a decision as to how we plan on implementing it and whether we want to rewrite analyzer to accommodate watched folders or just say that we no longer support this and provide a CLI tool for people to import files from their filesystem.

**Comment by frecuencialibre** [comment]
i think this discussion is now in #514. my memory is that following my testing on a full library we concluded that pynotify isn't the best choice. my vote is for a CLI tool. i also noticed recently that the script that @hairmare [mentions](https://github.com/LibreTime/libretime/pull/514#issuecomment-444171011) is [in the documentation](http://libretime.org/manual/automated-file-import/), so that should be updated if need be.

**Comment by paddatrapper** [comment]
Consideration of #845 needs to be taken into account when watch files are implemented. There could be issues with cache consistency

**Comment by kmahelona** [comment]
I've now encountered #845 and I'm wondering how I can fix this. It's causing major issues in production (mainly our news person spends a lot of time doing the news only to find that it doesn't get played properly because an old cache is being served). The md5hash was my immediate response but it seems that's not ideal?

**Comment by paddatrapper** [comment]
What shortcomings do you see with MD5sum checks?

**Comment by kmahelona** [comment]
Nothing aside from it takes longer to calculate. Something on order of hundredths of a second versus milliseconds for file size check (I did a quick test on a particular file that was about 7MB). But if file size is different would md5?

**Comment by paddatrapper** [comment]
If the file size is different, then the contents are different, so the MD5sum would be different (ignoring collisions). However, if the file size is the same the MD5sum is not necessarily the same.

**Comment by Robbt** [comment]
Hmm, I think a md5 check would be a good way of doing this, it only happens during pypo playout scheduling and so the delay shouldn't be negatively impactful considering items are copied over and scheduled only when they are played.

**Comment by paddatrapper** [comment]
Looking to duplicate the functionality of Airtime's Media Monitor
